# 4.5 Data Consistency Checks

##### This script contains the following points

#### 01. Importing libraries

#### 02 Importing data

#### 03 Mixed-type data

#### 04 Missing values

#### 05 Duplicates

#### 06 Task

# 01. Importing libraries

In [42]:
# importing libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [43]:
path = r'/Users/mainframe/Documents/Instacart Basket Analysis'

In [44]:
# Importing products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [45]:
# Importing orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03 Mixed-type data

In [46]:
# creating a dataframe
df_test = pd.DataFrame()

In [47]:
# creating a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [48]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [49]:
# checking for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


/var/folders/rg/l7v73y996hv36vxgfkdf06lc0000gp/T/ipykernel_14562/968655696.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [50]:
# changing mixed type to string
df_test['mix'] = df_test['mix'].astype('str')

# 04 Missing values

In [51]:
# checking for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [52]:
# creating subset of df of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [53]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [54]:
df_prods.shape

(49693, 5)

In [55]:
# creating subset of df of non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [56]:
df_prods_clean.shape

(49677, 5)

# 05 Duplicates

In [57]:
# checking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [58]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [59]:
# creating new df without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [60]:
df_prods_clean_no_dups.shape

(49672, 5)

In [61]:
# exporting cleaned dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 06 Task

In [62]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   hour_ordered  days_since_prior_order  
0             8                     NaN  
1             7                    15.0  
2            12                    21.0  
3             7                    29.0  
4            15                    28.0

In [63]:
# 2. running data consistency check on df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_ordered  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [64]:
# 3. checking for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/rg/l7v73y996hv36vxgfkdf06lc0000gp/T/ipykernel_14562/1693045285.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/rg/l7v73y996hv36vxgfkdf06lc0000gp/T/ipykernel_14562/1693045285.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/rg/l7v73y996hv36vxgfkdf06lc0000gp/T/ipykernel_14562/1693045285.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/rg/l7v73y996hv36vxgfkdf06lc0000gp/T/ipykernel_14562/1693045285.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

In [65]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
hour_ordered                int64
days_since_prior_order    float64
dtype: object

####  4. No mixed data types found

In [66]:
# 5. checking for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_ordered                   0
days_since_prior_order    206209
dtype: int64

In [67]:
df_ords.shape

(3421083, 7)

In [68]:
# finding median of days_since_prior_order
median_value = df_ords['days_since_prior_order'].median()

In [69]:
print('Median:', median_value)

Median: 7.0


#### the median days_since_prior_order is 7

In [70]:
# creating subset of missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [71]:
df_ords_nan.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
11          11   2168274        2             1                   2   
26          26   1374495        3             1                   1   
39          39   3343014        4             1                   6   
45          45   2717275        5             1                   3   

    hour_ordered  days_since_prior_order  
0              8                     NaN  
11            11                     NaN  
26            14                     NaN  
39            11                     NaN  
45            12                     NaN

In [72]:
df_ords_nan.tail()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         hour_ordered  days_since_prior_order  
3420930            12                     NaN  
3420934            18                     NaN  
3421002            19                     NaN  
3421019            15                     NaN  
3421069            11                     NaN

In [73]:
# 6. Addressing the missing values in df_ords

#### There are 206,209  missing values in the days_since_prior_order column. There are also 206,209 user ids. For every user id's first order, there is a missing value in the days_since_prior_order column. This is because there couldn't be an order prior to the user id's first order. In this case "NaN" is a suitable value.

In [74]:
# 7. checking df_ords for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [75]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, hour_ordered, days_since_prior_order]
Index: []

In [76]:
# 8. addressing duplicates

#### No duplicates were found.

In [77]:
# 9. The cleaned df_prods data was already exported.
# exporting df_ords
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))